up-subscription/src/usubscription.rs

```diff
@@ -0,0 +1,537 @@
+
+/// Whether to include 'up:'  uProtocol schema prefix in URIs in log and error messages
+pub const INCLUDE_SCHEMA: bool = false;
+
+// Remote-subscribe operation ttl; 5 minutes in milliseconds, as per https://github.com/eclipse-uprotocol/up-spec/tree/main/up-l3/usubscription/v3#6-timeout--retry-logic
+pub(crate) const UP_REMOTE_TTL: u32 = 300000;
+
+/// This trait (and the comprised UTransportHolder trait) is simply there to have a generic type that
+/// usubscription Listeners deal with, so that USubscriptionService can be properly mocked.
+pub trait USubscriptionServiceAbstract:
+    USubscription + LocalUriProvider + UTransportHolder
+{
+}
+
+/// This trait primarily serves to provide a hook-point for using the mockall crate, for mocking USubscriptionService objects
+/// where we also need/want to inject custom/mock UTransport implementations that subsequently get used in test cases.
+pub trait UTransportHolder {
+    fn get_transport(&self) -> Arc<dyn UTransport>;
+}
+
+impl UTransportHolder for USubscriptionService {
+    fn get_transport(&self) -> Arc<dyn UTransport> {
+        self.up_transport.clone()
+    }
+}
+
+/// Core landing point and coordination of business logic of the uProtocol USubscription service. This implementation usually would be
+/// front-ended by the various `listeners` to connect with corresponding uProtocol RPC server endpoints.
+///
+/// Functionally, the code in this context primarily cares about:
+/// - input validation
+/// - interaction with / orchestration of backends for managing subscriptions (`usubscription_manager.rs`) and dealing with notifications (`usubscription_notification.rs`)
+#[derive(Clone)]
+pub struct USubscriptionService {
+    pub name: String,
+    pub own_uri: UUri,
+
+    pub(crate) up_transport: Arc<dyn UTransport>,
+    subscription_sender: UnboundedSender<SubscriptionEvent>,
+    notification_sender: UnboundedSender<usubscription_notification::Event>,
+}
+
+/// Implementation of uProtocol L3 USubscription service
+impl USubscriptionService {
+    /// Construct a new USubscriptionService
+    /// Atm this will directly spin up two tasks which deal with subscription and notification management, with no further action
+    /// required, but also no explicit shutdown operation yet - that's a TODO.
+    ///
+    /// # Arguments
+    ///
+    /// * `name` - An optional name for the USubscription service, might be used in logging (TODO)
+    /// * `own_uri` - The UUri of this USUbscription service (setting a Authority might be sufficient - TODO)
+    /// * `up_transport` - Implementation of UTransport to be used by this USUbscription instance, for sending Listener-responses and Notifications
+    /// * `up_client` - Implementation of RpcClient to be used by this USUbscription instance, for performing remote-subscribe operations
+    ///
+    pub fn new(
+        name: Option<String>,
+        own_uri: UUri,
+        up_transport: Arc<dyn UTransport>,
```

Comment:
Yes, that is the idea - also, right now, for Reasons(TM), it would be possible to use different implementations for RpcClient vs UTransport. But I do expect this corner to change somewhat, when I pick up @sophokles73 recent changes around L2 implementations in up-rust.